runner/conmon_test/conmon_test.go

```diff
@@ -0,0 +1,201 @@
+package conmon_test
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/containers/conmon/runner/conmon"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var (
+	conmonPath   = "/usr/bin/conmon"
+	ctrID        = "abcdefghijklm"
+	validPath    = "/tmp"
+	validLogPath = "/tmp/log"
+	invalidPath  = "/not/a/path"
+)
+
+func getConmonOutputGivenOptions(options ...conmon.ConmonOption) (string, string) {
+	var stdout bytes.Buffer
+	var stderr bytes.Buffer
+
+	options = append(options, conmon.WithStdout(&stdout), conmon.WithStderr(&stderr))
+
+	ci, err := conmon.CreateAndExecConmon(options...)
+	Expect(err).To(BeNil())
+
+	ci.Wait()
+
+	return stdout.String(), stderr.String()
+}
+
+var _ = Describe("conmon", func() {
+	Describe("version", func() {
+		It("Should return conmon version", func() {
+			out, _ := getConmonOutputGivenOptions(
+				conmon.WithVersion(),
+				conmon.WithPath(conmonPath),
+			)
+			Expect(out).To(ContainSubstring("conmon version"))
+			Expect(out).To(ContainSubstring("commit"))
+		})
+	})
+	Describe("no container ID", func() {
+		It("should fail", func() {
+			_, err := getConmonOutputGivenOptions(
+				conmon.WithPath(conmonPath),
+			)
+			Expect(err).To(ContainSubstring("conmon: Container ID not provided. Use --cid"))
+		})
+	})
+	Describe("no container UUID", func() {
+		It("should fail", func() {
+			_, err := getConmonOutputGivenOptions(
+				conmon.WithPath(conmonPath),
+				conmon.WithContainerID(ctrID),
+			)
+			Expect(err).To(ContainSubstring("Container UUID not provided. Use --cuuid"))
+		})
+	})
```

Comment:
I'm really happy to see unit-tests period, so I'm hesitant about suggesting you avoid (and/or remove) the negative cases.

My main reason against negative unit-tests is they can be deceptively complex (as proving a "negative" often is).  They also tend to be the most burdensome in terms of maintenance of test and subject code.  It's okay IMHO if unit tests are only positive and interface-validating.  We can rely on integration tests to catch the negative cases worth investigating.

Reply:
For the first set of options (the ones I've written tests for so far), conmon fails without them in the inital pid. These tests are all negative because they guarentee conmon requires them. The tests will get more complex from here. I don't mind the simple string matching as evidence of failure for now, and if proving "negative" gets overly burdensome in the future, I'll avoid them :smile:

Reply:
sounds good.